File: swift-sdk/Sources/GistAdsSDK/Services/AdAPIService.swift
//
//  AdAPIService.swift
//  GistAdsSDK
//
//  Service for communicating with Gist Ads API
//

import Foundation

/// API service for fetching ads from Gist Search API
class AdAPIService {
    
    private let baseURL: String
    private let publisherID: String
    private let publisherKey: String
    private let apiVersion: String
    
    init(baseURL: String, publisherID: String, publisherKey: String, apiVersion: String? = nil) {
        self.baseURL = baseURL
        self.publisherID = publisherID
        self.publisherKey = publisherKey
        self.apiVersion = apiVersion ?? APIConstants.apiVersion()
    }
    
    /// Fetch ads from the search API
    /// - Parameters:
    ///   - query: The search query text
    ///   - geo: Geographic location (e.g., "US", "GB")
    ///   - adTypes: Optional array of ad types to filter
    ///   - answer: Optional answer string for v2 (defaults to query if nil)
    /// - Returns: HTML string containing the ad iframe
    func fetchAd(query: String, geo: String, adTypes: [AdType]?, answer: String? = nil) async throws -> String {
        // Construct the API endpoint using URL(string:relativeTo:) for safer construction
        guard let base = URL(string: baseURL) else {
            throw AdAPIError.invalidURL
        }
        let endpoint = APIConstants.searchEndpoint(for: apiVersion)
        guard let url = URL(string: endpoint, relativeTo: base) else {
            throw AdAPIError.invalidURL
        }
        
        // Create request
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(publisherID, forHTTPHeaderField: "Publisher-ID")
        request.setValue(publisherKey, forHTTPHeaderField: "Publisher-Key")
        
        // Build request body using factory function
        let adTypeStrings = adTypes?.map { $0.rawValue }
        let searchRequest = createSearchRequest(
            version: apiVersion,
            query: query,
            geo: geo,
            adTypes: adTypeStrings,
            answer: answer
        )
        
        let requestBody = try JSONEncoder().encode(searchRequest)
        request.httpBody = requestBody
        
        // Make the request
        let (data, response) = try await URLSession.shared.data(for: request)
        
        // Check response status
        guard let httpResponse = response as? HTTPURLResponse else {
            throw AdAPIError.invalidResponse
        }
        
        guard httpResponse.statusCode == 200 else {
            throw AdAPIError.httpError(statusCode: httpResponse.statusCode, response: data)
        }
        
        // Handle response based on API version
        return try handleResponse(version: apiVersion, data: data)
    }
    
    /// Handle API response based on version
    /// - Parameters:
    ///   - version: API version string
    ///   - data: Response data
    /// - Returns: HTML string containing the ad
    private func handleResponse(version: String, data: Data) throws -> String {
        // Both v1 and v2 use the same JSON format with selection array
        return try parseJSONResponse(data: data)
    }
    
    /// Parse JSON response and extract iframe URL
    /// - Parameter data: Response data containing JSON
    /// - Returns: HTML string containing the ad iframe
    private func parseJSONResponse(data: Data) throws -> String {
        let searchResponse: SearchResponse
        do {
            searchResponse = try JSONDecoder().decode(SearchResponse.self, from: data)
        } catch {
            throw AdAPIError.invalidData(underlying: error)
        }
        
        // Check if we have a selection with ads
        guard let selection = searchResponse.selection, !selection.isEmpty else {
            throw AdAPIError.noAdsAvailable
        }
        
        // Extract iframeUrl from first selection item
        guard let firstAd = selection.first,
              let iframeUrl = firstAd.iframeUrl,
              !iframeUrl.isEmpty else {
            throw AdAPIError.missingIframeUrl
        }

        // Generate iframe HTML using utility
        return IframeHTMLGenerator.generate(iframeUrl: iframeUrl)
    }
}

/// Errors that can occur during API operations
enum AdAPIError: LocalizedError {
    case invalidURL
    case invalidResponse
    case invalidData(underlying: Error)
    case httpError(statusCode: Int, response: Data?)
    case noAdsAvailable
    case missingIframeUrl
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid API URL"
        case .invalidResponse:
            return "Invalid response from server"
        case .invalidData(let underlying):
            return "Unable to parse response data: \(underlying.localizedDescription)"
        case .httpError(let statusCode, _):
            return "HTTP error: \(statusCode)"
        case .noAdsAvailable:
            return "No ads available for this query"
        case .missingIframeUrl:
            return "Ad response missing iframe URL"
        }
    }
}

File: android-sdk/src/main/java/com/gist/ads/sdk/services/AdAPIService.kt
package com.gist.ads.sdk.services

import com.gist.ads.sdk.APIConstants
import com.gist.ads.sdk.models.AdType
import com.gist.ads.sdk.models.SearchResponse
import com.gist.ads.sdk.models.createSearchRequest
import com.gist.ads.sdk.utils.IframeHTMLGenerator
import com.google.gson.Gson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * API service for fetching ads from Gist Search API
 */
class AdAPIService(
    private val baseUrl: String,
    private val publisherId: String,
    private val publisherKey: String,
    private val apiVersion: String = APIConstants.defaultApiVersion(),
    enableLogging: Boolean = false
) {
    private val client: OkHttpClient
    private val gson = Gson()
    
    init {
        val builder = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
        
        if (enableLogging) {
            val loggingInterceptor = HttpLoggingInterceptor().apply {
                level = HttpLoggingInterceptor.Level.BODY
            }
            builder.addInterceptor(loggingInterceptor)
        }
        
        client = builder.build()
    }
    
    /**
     * Fetch ads from the search API
     * @param query The search query text
     * @param geo Geographic location (e.g., "US", "GB")
     * @param adTypes Optional list of ad types to filter
     * @param answer Optional answer string for v2 (defaults to query if null)
     * @param theme Theme preference - "light" or "dark" (resolved from system if "system" was selected)
     * @return HTML string containing the ad iframe
     * @throws AdAPIException if the request fails
     */
    suspend fun fetchAd(
        query: String,
        geo: String,
        adTypes: List<AdType>?,
        answer: String? = null,
        theme: String
    ): String = withContext(Dispatchers.IO) {
        try {
            // Build request body using factory function
            val adTypeStrings = adTypes?.map { it.value }
            val searchRequest = createSearchRequest(
                version = apiVersion,
                query = query,
                geo = geo,
                adTypes = adTypeStrings,
                answer = answer
            )
            
            val jsonBody = gson.toJson(searchRequest)
            val requestBody = jsonBody.toRequestBody("application/json".toMediaType())
            
            // Build HTTP request with dynamic endpoint
            val endpoint = "$baseUrl${APIConstants.searchEndpoint(apiVersion)}"
            val request = Request.Builder()
                .url(endpoint)
                .post(requestBody)
                .addHeader("Content-Type", "application/json")
                .addHeader("Publisher-ID", publisherId)
                .addHeader("Publisher-Key", publisherKey)
                .build()
            
            // Execute request
            val response = client.newCall(request).execute()
            
            if (!response.isSuccessful) {
                throw AdAPIException.HttpError(response.code)
            }
            
            val responseBody = response.body?.string()
                ?: throw AdAPIException.InvalidData
            
            // Parse JSON response and extract iframe URL
            parseJSONResponse(responseBody, theme)
            
        } catch (e: IOException) {
            throw AdAPIException.NetworkError(e)
        } catch (e: AdAPIException) {
            throw e
        } catch (e: Exception) {
            throw AdAPIException.UnknownError(e)
        }
    }
    
    /**
     * Parse JSON response and extract iframe URL
     * @param data Response data containing JSON
     * @param theme Theme preference for iframe
     * @return HTML string containing the ad iframe
     */
    private fun parseJSONResponse(data: String, theme: String): String {
        val searchResponse: SearchResponse
        try {
            searchResponse = gson.fromJson(data, SearchResponse::class.java)
        } catch (e: Exception) {
            throw AdAPIException.InvalidData
        }
        
        // Check if we have a selection with ads
        val selection = searchResponse.selection
        if (selection.isNullOrEmpty()) {
            throw AdAPIException.NoAdsAvailable
        }
        
        // Extract iframeUrl from first selection item
        val firstAd = selection.first()
        val iframeUrl = firstAd.iframeUrl
        
        if (iframeUrl.isNullOrEmpty()) {
            throw AdAPIException.MissingIframeUrl
        }
        
        // Generate iframe HTML using utility with theme
        return IframeHTMLGenerator.generate(iframeUrl, theme)
    }
}

/**
 * Errors that can occur during API operations
 */
sealed class AdAPIException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    object InvalidUrl : AdAPIException("Invalid API URL configured")
    object InvalidResponse : AdAPIException("Invalid response from server")
    object InvalidData : AdAPIException("Invalid response format: Unable to parse server response")
    
    data class HttpError(val statusCode: Int) : AdAPIException(getHttpErrorMessage(statusCode)) {
        companion object {
            private fun getHttpErrorMessage(code: Int): String = when (code) {
                401 -> "Authentication failed: Invalid publisher credentials"
                403 -> "Access forbidden: Check your publisher permissions"
                404 -> "API endpoint not found: Verify SDK version compatibility"
                429 -> "Rate limit exceeded: Too many requests"
                500 -> "Server error: Please try again later"
                503 -> "Service temporarily unavailable"
                else -> "HTTP error $code: Request failed"
            }
        }
    }
    
    object NoAdsAvailable : AdAPIException("No ads available: No matching ads found for this query and location")
    object MissingIframeUrl : AdAPIException("Invalid ad response: Missing iframe URL from server")
    
    data class NetworkError(val error: Throwable) : AdAPIException(
        "Network error: ${error.message ?: "Unable to connect to ad server"}",
        error
    )
    
    data class UnknownError(val error: Throwable) : AdAPIException(
        "Unexpected error: ${error.message ?: "Unknown error occurred"}",
        error
    )
}


